### DISCUSSION TITLE
Full stack folder structure

### DISCUSSION THREAD

**Discussion by nimonian** [swe-delivery/swe-tube#2] [General]
In a full stack bootcamp project, how should we encourage apprentices to organise their services with respect to repos?
Do we have
 1. one repo, a folder called server for the back-end and a folder called client  for the front-end
 1. two repos, one for the front-end and one for the back-end
 1. something different
?

**Comment by nimonian** [comment]
### Copy of the discussion so far

Simon Haines
  [2 hours ago](https://multiverseio.slack.com/archives/C04B62EEDLK/p1675871485242719?thread_ts=1675870930.002139&cid=C04B62EEDLK)
I have similar feelings - how to make it digestible and convenient for newbies



Jordan McNamara
  [2 hours ago](https://multiverseio.slack.com/archives/C04B62EEDLK/p1675871560488669?thread_ts=1675870930.002139&cid=C04B62EEDLK)
I could go both ways on this, to be honest.  Mono means less context switching, maybe less git issues? It also means two package.json files and node_modules


Jordan McNamara
  [2 hours ago](https://multiverseio.slack.com/archives/C04B62EEDLK/p1675871690901469?thread_ts=1675870930.002139&cid=C04B62EEDLK)
We may not get this right, but in terms of "best practice" we are optimising for our context and not necessarily for other concerns.
I think we then avoid the trap of being caught in the arbiters of "best practice."  Especially if we expand our bootcamp stack to include other languages.


Simon Haines
  [2 hours ago](https://multiverseio.slack.com/archives/C04B62EEDLK/p1675871823843349?thread_ts=1675870930.002139&cid=C04B62EEDLK)
Yes I like (a) for those reasons and it is what I've gravitated towards in the past
I also found it makes it easier to incorporate docker-compose in the future
It would probably make it easier for teams to collaborate too (if there are 2 front-end and 2 back-end devs for example), less git issues as you say
So long as they don't npm i express in the root of the project :smile: (edited) 


Simon Haines
  [2 hours ago](https://multiverseio.slack.com/archives/C04B62EEDLK/p1675871956836449?thread_ts=1675870930.002139&cid=C04B62EEDLK)
and yes I agree we need to strike a balance between "industrial realism" and "breaking it down for newcomers"


Jordan McNamara
  [2 hours ago](https://multiverseio.slack.com/archives/C04B62EEDLK/p1675872139648239?thread_ts=1675870930.002139&cid=C04B62EEDLK)
If anyone tells me a universal rule, I'm immediately sceptical!

**Comment by accuser** [comment]
I would typically prefer a monorepo using [PNPM](https://pnpm.io) and [Turbo](https://turbo.build/repo) but neither of these are used in bootcamp. Turbo is easy to introduce, but PNPM is a replacement for NPM and requires a little more attention. (Why a monorepo? There are a lot of arguments for and against, but I fall on the 'we are building a whole system' side, and the system is maintained in a single repo.) 

This actually presents an opportunity for additional content to be delivered in Community to a) introduce additional tools (and languages and frameworks) and b) consider how we make choices and mitigate the consequences.

**Reply by Jaardy** [reply]
Upvote for Community content!

**Comment by accuser** [comment]
I think about deployment in the organisation of the repo: what are the deployable units, what do they depend on, and where are they being deployed? 

Informed by Turbo and [Netlify](https://netlify.com) I'll organise the repo into deployable apps and sharable packages (which may or may not be deployed).

For a project like this I'd probably be thinking of apps like:

- `api` wrapping all data and 3rd-party services
- `web` the 'marketing' website which is served to unauthenticated visitors
- `app` the application for authenticated users
- `cli` for interacting with `api` (useful for admin and e2e testing)

I'd also be thinking of packages like:

- `ui` for the design system, using [Storybook](https://storybook.js.org)
- `content` for [CMS](https://sanity.io)
- `data` for the data schema

**Reply by accuser** [reply]
Netlify also offers features for serverless functions (including scheduled), edge functions, and on-demand builders (think: cached). Beyond bootcamp, but certainly not out of the realms of the back-end module, nor Community.

**Reply by accuser** [reply]
CI/CD is also more practical with a monorepo: if the system is atomic, validation to deployment is straightforward.

**Reply by nimonian** [reply]
I love the idea of thinking in terms of deployable apps and shared packages, it helps apprentices realise which part of the stack they're working on, each folder "lives" in a place on the internet, it's almost visual.

Having an argued, rationalised approach to project organisation will help so much when we're going through part 2 of bootcamp - we can explain what each part of the project does and how it works together up-front so apprentices have the full picture of what they're aiming for.

**Comment by accuser** [comment]
I'm thinking that this project is going to be useful for bootcampers to actually use, but we're also pulling back the curtain so that they can see the decision we made about the architecture, design, and implementation. We will likely need a meta-project that presents this in an accessible way, which should probably include documentation and articles.

**Reply by accuser** [reply]
We could add `doc` to the list of apps to host this.

**Reply by nimonian** [reply]
So much this, incredible idea

**Reply by Jaardy** [reply]
An idea introduced by Andy that might build on this idea of a live code base:  Apprentice contributions.  In order for it to be repeatable, the suggestion was to have a live code base available as a template and then giving them a "ticket".  It is a bit of an immersive experience but still "safe".

**Reply by nimonian** [reply]
This kind of "immersive experience" is what I've been referring to as "embedded durable skills" and I am _all for it_. When giving the apprentices tasks, how could we emulate the way tasks are communicated, tracked and evaluated in industry?

Matthew suggested scribbling a spec document on a beer mat and mailing it to them :rofl: I am genuinely tempted...

**Reply by accuser** [reply]
It might be worth getting the 'vanilla' branch to a point where it is useful, and tagging it. We could then start each bootcamp with a fresh 'vanilla' branch, against which we could raise tickets for features, bug fixes, documentation, etc.
